base_layer/key_manager/src/cipher_seed.rs

```diff
@@ -155,223 +155,205 @@ impl CipherSeed {
         }
     }
 
+    /// Generate an encrypted seed from a passphrase
     pub fn encipher(&self, passphrase: Option<String>) -> Result<Vec<u8>, KeyManagerError> {
-        let mut plaintext = self.birthday.to_le_bytes().to_vec();
-        plaintext.append(&mut self.entropy().clone().to_vec());
-
-        let passphrase = passphrase.unwrap_or_else(|| DEFAULT_CIPHER_SEED_PASSPHRASE.to_string());
+        // Derive encryption and MAC keys from passphrase and main salt
+        let passphrase = Zeroizing::new(passphrase.unwrap_or_else(|| DEFAULT_CIPHER_SEED_PASSPHRASE.to_string()));
+        let (encryption_key, mac_key) = Self::derive_keys(&passphrase, &self.salt)?;
 
-        // generate the current MAC
+        // Generate the MAC
         let mut mac = Self::generate_mac(
             &self.birthday.to_le_bytes(),
-            &self.entropy(),
-            &[CIPHER_SEED_VERSION],
+            self.entropy.as_ref(),
+            CIPHER_SEED_VERSION,
             &self.salt,
-            passphrase.as_str(),
+            mac_key.as_ref(),
         )?;
 
-        plaintext.append(&mut mac);
+        // Assemble the secret data to be encrypted: birthday, entropy, MAC
+        let mut secret_data = Zeroizing::new(Vec::<u8>::with_capacity(
+            CIPHER_SEED_BIRTHDAY_BYTES + CIPHER_SEED_ENTROPY_BYTES + CIPHER_SEED_MAC_BYTES,
+        ));
+        secret_data.append(&mut self.birthday.to_le_bytes().to_vec());
+        secret_data.append(&mut self.entropy.clone());
```

Comment:
self.entropy.clone()  leaks (I believe the compiler won't optimise the clone out because cloned value is being mutated/emptied but not zeroed)
```suggestion
        secret_data.extend(&self.birthday.to_le_bytes());
        secret_data.extend(&self.entropy);
```